### DISCUSSION TITLE
Nix alternative (again)

### DISCUSSION THREAD

**Discussion by SR-G** [srid/neuron#622] [Installation]
So i think this is a recurrent topic ... but what is the current status of neuron as a binary execution / outside nix ?

It seems that : 
- regular binaries where considered (#183)
- Nix alternative were considered (#243, with pre-requesites done in the meanwhile), also related to #417
- but then have been discontinued (if i'm not mistaken)

In my situation, i have an Obsidian set of .md that i would like to convert to HTML. Neuron seems perfect for this (about the workflow it would allow + the final result). I'm running windows (WSL), but sadly can't use the docker image. So i'm just in need of a statically compiled binary (with glibc).

So, isn't it possible to just have either again static binaries (the latest one released here on github is quite old), either another package manager than Nix ?

Thanks in advance for any help.

**Comment by srid** [comment]
@SR-G The static builds are in broken state, pending someone volunteering to fix it.

If you already have WSL (WSL 2, I hope), then why not [install it via Nix](https://neuron.zettel.page/install) since Nix is supported on WSL?

**Reply by SR-G** [reply]
Well, it's the first time i encounter the NIX ecosystem, and to be honest, it seems way too intrusive and not really flexible (system-wide installation in /nix, grabs a lot of files (i tryed in a VM on another server = 6.2GB of files for a 30mb binary, which is just insane - and not possible on my target small windows laptop), ... + i wondered if i could maybe use a docker container in another linux server to somehow build the binary from sources (in a very different way than using the neutron docker image = rather to try to build the neutron binary in another way), and it seems far away from possible (root installation of nix not possible inside a docker container (whereas it's using sudo ... sigh), nothing available out of the box to produce a binary (as far as it seems, but maybe i've missed it), ...))

So to be honest, i'm not impressed at all by this package manager at first glance - of course, this is a personal opinion and i understand others may think differently. 

Hence the question about having a way more standard way (in addition to docker) to be able to execute Neutron (like pre-compiled binaries).

**Reply by srid** [reply]
Well, one option is to try https://github.com/srid/nix-bundle/tree/neuron-stable - which does provide a bundle <100M that can be copied to any Linux system and ran independently.

But yes ideally having static binaries would be better. Running `nix-build -j auto static.nix` in the neuron Git repo is supposed to build it, if anyone is looking to contribute fixes to the static build.

**Reply by SR-G** [reply]
I just tried nix-bundle, but i'm unable to do anything with it, running `./nix-bundle.sh neutron /bin/neutron` leads to 
`error: undefined variable 'neutron' at (string):1:82`

Guess i have to stop investigating the neutron way, the ecosystem is way too complex for me.

Thanks for your answers anyway.

**Reply by bew** [reply]
@SR-G the project is named `neuron`, not `neutron` :eyes:

**Reply by SR-G** [reply]
Correct.
I did a few tryes, with different errors, with proper names it's giving me : 
`error: getting attributes of path '/home/bin/neuron': No such file or directory`
(i assume that first parameter is the package to be built, second one is the target filename to be generated, so obviously it's not expected to be available before the command has been executed - but the documentation is anyway quite confusing about this second parameter "[...] while /bin/hello indicates the path of the executable _relative to hello that you want to run_" (for the "hello" example)

**Reply by bew** [reply]
> while /bin/hello indicates the path of the executable relative to hello that you want to run

AFAIK this means that **_in_** the package 'hello' (that is built by nix-bundle by running `nix build` or sth similar), it will take the binary at `bin/hello` (the leading `/` representing the root of the generated package).

You said earlier that you had to download 6G of data to build neuron, I suspect that it didn't found binary caches and tried to recompile the world..
Could you paste the output of `nix-channel --list` ? (maybe with sudo if that commands doesn't print anything, meaning you installed Nix system-wide (with a daemon) and not user-wide)

**Reply by SR-G** [reply]
With root is doesn't print anything, with current user : 
`nixpkgs https://nixos.org/channels/nixpkgs-unstable`

Of course i have the `neuron` binary inside nix cache : 

```
whereis neuron
neuron: /nix/store/s2kc0agagr35f70dqkzpar3kjdhn5mm5-user-environment/bin/neuron
```

Indeed based on the /nix/* content, it has downloaded the world (glibc, gcc, bash, binutils, gtk, ...)

**Reply by srid** [reply]
`/bin/neuron` looks correct - but I think the first argument should be the nix expression that uses the default.nix in this repo. See the removed commands in this commit for a working version:

https://github.com/srid/neuron/commit/1d5cbba1667565229ab99278c13709d2200890be#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133dd

**Reply by bew** [reply]
@srid FYI I tried to build it statically to test what you meant by 'it-needs-fixes', but as it tried to recompile the world statically I stopped it. I don't want to build the world, then have something break at the end and require me to recompile everything again..
Could you give bit more info about what's broken / at which stage it doesn't work / etc? (And make an issue, so you/we can track this :smiley:)

**Reply by srid** [reply]
Well, it should compile the world only the first time - then the packages are cached, and when you re-run nix-build it should reuse the cache. I usually start leaving it running overnight and then check again in the morning :-)

Looks to be https://github.com/NixOS/nixpkgs/issues/118731

![image](https://user-images.githubusercontent.com/3998/119282567-1c91a780-bc08-11eb-8365-b2eb90b888f6.png)

**Comment by ouuan** [comment]
Just curious, @srid how large is your `/nix` directory? Are you using NixOS?

I'm using Nix for nothing but neuron, and now my `/nix` is 9GiB, which is half of everything in my root directory except `/home` and `/nix`. Run only `nix-env -if` to install neuron takes 6GiB. Executing `bin/run` & trying to build the static binary file and then aborting the build take more.

**Reply by srid** [reply]
Yup, this is why we need https://github.com/srid/neuron/issues/626 (then you won't need nix as an user)

Generally speaking, to derive most value from Nix - you would be using it for many other packages as well. This is what I do on Windows (when on WSL2) or Linux (via NixOS). Not what you asked for, but you can try NixOS in lieu of your current $distro to see what difference that makes (virtually nothing in /usr, due to everything being shared under /nix).
